### PR TITLE
test verifying upgraded backend

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Adds unit tests for upgrade error handling

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @cjcullen

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go [hunk 1]
```diff
@@ -538,6 +539,68 @@ func TestProxyUpgradeErrorResponse(t *testing.T) {
 	assert.Contains(t, string(msg), expectedErr.Error())
 }
 
+func TestProxyUpgradeErrorResponseTerminates(t *testing.T) {
+	for _, intercept := range []bool{true, false} {
+		for _, code := range []int{200, 400, 500} {
```
Comment (cheftako) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: Can we add an invalid response code (eg 0, -1) to test that we handle more serious errors as well? I know they should not occur but I believe we have seen bugs where those response codes occur at least in our integration tests.

Comment (cheftako) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: I think it allows the test to protect us from an invalid (future) assumption about only having to handle "valid" response codes.

Comment (cheftako) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: Tempted to suggest adding a test for something like an error being thrown in the handle case.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: sending 0 or -1 isn't even possible using standard response writers. this test case is specifically around how the proxy handles well-formed responses from the backend that are not upgrade responses. error responses like the ones you describe short-circuit earlier. if we want to try to trigger testing those with direct connection bit writing, we can, but I'd like to do that in a different test

Comment (cheftako) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: Fair enough. Happy to lgtm as this only makes things better. However we know from our current run of flaky tests that we sometimes return a HTTP response code of 0.

**Pull request merged**